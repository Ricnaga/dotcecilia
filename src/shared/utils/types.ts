export type OnlyBooleanKeys<T extends object> = {
  [K in keyof T]-?: T[K] extends boolean ? K : never;
}[keyof T];

export type CalculatorType<T extends object> = Record<'values', T> & {
  onChange: (key: keyof T, value: number | string) => void;
  onChecked?: (key: OnlyBooleanKeys<T>) => void;
  onPrint: () => void;
};
