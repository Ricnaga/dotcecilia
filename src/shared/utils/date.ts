type WorkDaysInMonthType = Record<'year' | 'month' | 'day', number>;

export const getWorkDaysInMonth = ({ day, month, year }: WorkDaysInMonthType) =>
  new Array<number>(day)
    .fill(Math.round(Math.random()))
    .reduce((accumulator, __, index) => {
      const weekDay = new Date(year, month - 1, index + 1).getDay();
      if (weekDay !== 0 && weekDay !== 6) return accumulator + 1;

      return accumulator;
    }, 0);
