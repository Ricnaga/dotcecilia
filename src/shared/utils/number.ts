export function convertToCurrencyFloat(value: number) {
  const numbersString = value
    .toString()
    .replace(/[^0-9]/g, '')
    .padStart(3, '0');
  const floatString = numbersString.replace(/^([0-9]*?)([0-9]{2})$/, '$1.$2');
  return parseFloat(floatString);
}

export const convertToBRL = (num: number, currency = true) => {
  const formattedBRL = new Intl.NumberFormat('pt-br', {
    style: 'currency',
    currency: 'BRL',
  }).format(num);

  if (!currency) return formattedBRL.split('R$').at(1);

  return formattedBRL;
};

export const currencyStringToFloat = (currencyString: string) =>
  parseFloat(currencyString);
