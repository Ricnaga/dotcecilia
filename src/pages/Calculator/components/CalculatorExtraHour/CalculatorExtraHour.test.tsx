import { OnlyBooleanKeys } from '@shared/utils/types';
import {
  render,
  fireEvent,
  screen,
  ByRoleMatcher,
  within,
} from '@testing-library/react';
import {
  CalculatorExtraHour,
  CalculatorExtraHourData,
} from './CalculatorExtraHour';

type ButtonsType = Array<
  {
    role: ByRoleMatcher;
    name: RegExp;
  } & Partial<{ value: number }>
>;

let mockExtraHour: CalculatorExtraHourData = {
  agreement: true,
  hours: 0,
  minimumWage: 0,
  salary: 0,
  unsanitary: true,
  workedMonths: 0,
};

let containerComponent: HTMLElement;

describe('Component: CalculatorExtraHour', () => {
  beforeEach(() => {
    const { container } = render(
      <CalculatorExtraHour
        extraHour={mockExtraHour}
        onChangeExtraHour={jest.fn(
          (key: keyof CalculatorExtraHourData, value: string | number) => {
            mockExtraHour = { ...mockExtraHour, [key]: value };
          },
        )}
        onCheckedExtraHour={jest.fn(
          (key: OnlyBooleanKeys<CalculatorExtraHourData>) => {
            mockExtraHour = { ...mockExtraHour, [key]: !mockExtraHour[key] };
          },
        )}
      />,
    );
    fireEvent.click(
      screen.getByRole('button', {
        name: /hora extra/i,
      }),
    );
    containerComponent = container;
  });

  it('should render correctly', () => {
    expect(containerComponent).toMatchSnapshot();
  });

  it('should test number inputs', () => {
    const spinButtons: ButtonsType = [
      {
        name: /salário/i,
        role: 'spinbutton',
        value: 1000,
      },
      {
        name: /horas/i,
        role: 'spinbutton',
        value: 10,
      },
      {
        name: /valor mínimo/i,
        role: 'spinbutton',
        value: 1000,
      },
      {
        name: /meses trabalhados/i,
        role: 'spinbutton',
        value: 10,
      },
    ];

    spinButtons.map((buttons) =>
      fireEvent.change(
        within(screen.getByText(buttons.name)).getByRole(buttons.role),
        {
          target: { valueAsNumber: buttons.value },
        },
      ),
    );

    expect(mockExtraHour.salary).toBe(1000);
    expect(mockExtraHour.hours).toBe(10);
    expect(mockExtraHour.minimumWage).toBe(1000);
    expect(mockExtraHour.workedMonths).toBe(10);
  });

  it('should test switch buttons', () => {
    const switchButtons: ButtonsType = [
      {
        role: 'switch',
        name: /insalubridade/i,
      },
      {
        role: 'switch',
        name: /acerto/i,
      },
    ];

    switchButtons.map((buttons) =>
      fireEvent.click(
        screen.getByRole(buttons.role, {
          name: buttons.name,
        }),
      ),
    );
    expect(mockExtraHour.unsanitary).toBeFalsy();
    expect(mockExtraHour.agreement).toBeFalsy();
  });
});
