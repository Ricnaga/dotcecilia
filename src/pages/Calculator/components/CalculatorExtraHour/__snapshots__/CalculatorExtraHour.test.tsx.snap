// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CalculatorExtraHour should render correctly 1`] = `
<div>
  <button
    aria-controls="headlessui-disclosure-panel-:r1:"
    aria-expanded="true"
    data-headlessui-state="open"
    id="headlessui-disclosure-button-:r0:"
    type="button"
  >
    <p
      class="accordion-title"
    >
      Hora extra
    </p>
  </button>
  <div
    class="transition duration-100 ease-out transform scale-95 opacity-0"
  >
    <div
      data-headlessui-state="open"
      id="headlessui-disclosure-panel-:r1:"
    >
      <div
        class="accordion-content"
      >
        <div
          class="col-span-3"
        >
          <div
            class="flex items-center"
          >
            <label
              class="mr-4"
              id="headlessui-label-:r2:"
            >
              Insalubridade
            </label>
            <button
              aria-checked="true"
              aria-labelledby="headlessui-label-:r2:"
              class="btn-switch bg-teal-500 "
              data-headlessui-state="checked"
              id="headlessui-switch-:r3:"
              role="switch"
              tabindex="0"
              type="button"
            >
              <span
                class="translate-x-6 inline-block h-4 w-4 transform rounded-full bg-white transition-transform"
              />
            </button>
          </div>
          <label
            class="text-2xl"
            for="number"
          >
            Salário
            <input
              min="0.00"
              placeholder="R$"
              step="0.01"
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          class="col-span-3"
        >
          <div
            class="flex items-center"
          >
            <label
              class="mr-4"
              id="headlessui-label-:r4:"
            >
              Acerto
            </label>
            <button
              aria-checked="true"
              aria-labelledby="headlessui-label-:r4:"
              class="btn-switch bg-teal-500 "
              data-headlessui-state="checked"
              id="headlessui-switch-:r5:"
              role="switch"
              tabindex="0"
              type="button"
            >
              <span
                class="translate-x-6 inline-block h-4 w-4 transform rounded-full bg-white transition-transform"
              />
            </button>
          </div>
          <label
            class="text-2xl"
            for="number"
          >
            Horas
            <input
              min="0.00"
              placeholder="Horas"
              step="0.01"
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          class="col-span-3 grid items-end h-full"
        >
          <label
            class="text-2xl"
            for="number"
          >
            Valor mínimo
            <input
              min="0.00"
              placeholder="R$ 1200,00"
              step="0.01"
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          class="col-span-3 grid items-end h-full"
        >
          <label
            class="text-2xl"
            for="number"
          >
            Meses trabalhados
            <input
              min="0.00"
              placeholder="Meses"
              step="0.01"
              type="number"
              value=""
            />
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
